### PR TITLE
feat: level8 support

### PR DESCRIPTION
**Pull request by seriousme**
Migration to Level 8.0.0 while still  supporting Level 6.0.1 and 7.0.1 as discussed in #27

### CODE REVIEW

**Comment by seriousme** [comment]
As discussed this is a work in progress ;-)

**Comment by robertsLando** [comment]
@mcollina I would like to know your opinion here. Does it make sense to support all those level versions or should we drop old ones and keep only latest maintained?

Also how does iterators  impact in performances?

**Comment by seriousme** [comment]
> Also how does iterators impact in performances?
Level mentioned in 7.0.1 that its streams were based on underlying iterators ;-)

The current idea is to shift the streams more to the edge of `aedes-persistence-level`. If a next version of Aedes would switch to iterators as well we could ditch streams all together ;-)

**Comment by robertsLando** [comment]
> if a next version of Aedes would switch to iterators as well we could ditch streams all together

I'm not a fun of iterators as I think them make the code less readable. BTW Matteo has a deep knowladge of nodejs core so I think could tell us if them are more or less (or equal) to streams in performances. IMO streams are the best ones as the backpressue it's handled internally by nodejs

**Comment by seriousme** [comment]
> > if a next version of Aedes would switch to iterators as well we could ditch streams all together
> 
> I'm not a fun of iterators as I think them make the code less readable. BTW Matteo has a deep knowladge of nodejs core so I think could tell us if them are more or less (or equal) to streams in performances. IMO streams are the best ones as the backpressue it's handled internally by nodejs

Funny enough I find that the iterator approach is more readable to me (personal pref ;-)) than the streams approach. 
Last commit has removed all the internal streams handling so you can check for yourself which one you like better.
The streams handlers that got removed (callback-stream, though2 etc) are all third party dependencies and some have not been updated in 6 years or so. That does not need to be a bad thing, but feels a bit uncomfy to me ;-)

Happy to undo last commit if you all feel that we should stick streams.

**Comment by seriousme** [comment]
As to the backpressure concerns: 

A generator works as a paused stream, only doing work when the next function in line requests a value.
So if you only take 5 items out of an iterable then the chain of generators that feed that iterable will only have processed 5 items ;-)

Wrt readability:

```javascript
async function * willsByBrokers (db, brokers) {
  for await (const blob of dbValues(db, WILL)) {
    const chunk = msgpack.decode(blob)
    if (!brokers) {
      yield chunk
    } else {
      if (!brokers[chunk.brokerId]) {
        yield chunk
      }
    }
  }
}
```
combined with:
```javascript
 streamWill (brokers) {
    return Readable.from(willsByBrokers(this._db, brokers))
  }
```
does the same as:
```javascript
LevelPersistence.prototype.streamWill = function (brokers) {
  const valueStream = createValueStream(this._db, WILL)

  if (!brokers) {
    return pump(valueStream, through.obj(function (blob, enc, cb) {
      cb(null, msgpack.decode(blob))
    }))
  }

  return pump(valueStream, through.obj(function (blob, enc, cb) {
    const chunk = msgpack.decode(blob)
    if (!brokers[chunk.brokerId]) {
      this.push(chunk)
    }
    cb()
  }))
}
```
It took me some time to figure out that `this.push(chunk)` basically meant `yield chunk` ;-)
(and yes I do know that if you use `through2` on a daily basis this is probably well known, however it is specific for `through2` as its predecessor `through` uses `this.queue(chunk)` 

Another example:  multiplexing generators does not require a 9.3Kb  [external module](https://www.npmjs.com/package/multistream), this is it:  ;-)
```javascript
async function * multiIterables (iterables) {
  for (const iter of iterables) {
    yield * iter
  }
}
```

Not trying to push you into anything here, just trying to show the options :-)

Cheers,
Hans

**Comment by robertsLando** [comment]
I agree with you, as I said I have not a lot of experience using iterators but in this example everything looks better 👍🏼 

Let's see what @mcollina thinks about this.

Is you PR ready for review?

**Comment by seriousme** [comment]
> Is you PR ready for review?

Yes, I'd like to think so :-)

It is functional complete and I tested it against Level 6.0.1, 7.0.1 and 8.0.0
Package.json now contains an engines field to warn against pre-14 installs
Removing `through2` and `concat-stream` triggered a bug in `aedes-persistance`,although `aedes-persistence-level` does not use these modules they need to be installed to be able to test.
```
Error: Cannot find module 'concat-stream'
Require stack:
- /homedir/aedes-persistence-level/node_modules/aedes-persistence/abstract.js
```
We might want to move those in package.json of `aedes-persistance` from `devDependencies` to `dependencies`, once that is done they can be removed here ;-)

We could add, as usual ;-), some more ideas:

- [ ] add typescript defs
- [ ] add testing in CI for the various Level versions (if we need to still support those ;-))
- [ ] generalize the db.get, db.put so we don't have to have to specify encoding and msgpack in every functional call
- [ ] migrate internals from callback style to async/await to prepare for the future (Level can work with both styles)
- [ ] create a separate "LevelPersistenceClassic" class with the streams and callbacks on top of a new "LevelPersistence" class that exposes generators and async/await style so any future version of Aedes can switch from the classic interface to the modern interface (but only if Matteo sees this as the way forward too ;-))

But then again you might want to see separate PR's for these, just let me know :-)

**Comment by robertsLando** [comment]
>  generalize the db.get, db.put so we don't have to have to specify encoding and msgpack in every functional call

I would do this for now, I like all your idea except the last one that would require a huge rewrite of also other persistences like mongodb and redis included aedes itself.

About the types I wish to move aedes to typescript one day but my free time in last months have been really low and also there I dunno if Matteo would agree with that or just prefers types definitions

**Comment by seriousme** [comment]
>I would do this for now, I like all your idea except the last one that would require a huge rewrite of also other persistences like mongodb and redis included aedes itself.

One could stick to the current API indefinitly, however it would not be to hard to create 2 shims:
- classic2modern (as mentioned above) to facilitate current Aedes to continue to use the current API while persistences migrate to a more modern API
- modern2classic to facilitate Aedes to use persistences that use the currrent API while Aedes has already migrated to a more modern API

However that would only make sense when there is benefit in moving from streams to iterators and from callbacks to async/await.

> About the types I wish to move aedes to typescript one day but my free time in last months have been really low and also there I dunno if Matteo would agree with that or just prefers types definitions

Just adding the type definitions without going full typescript would help implementers of other persistences, however it might be more convenient to start that with `aedes-persistence` first as that is more or less the golden standard ;-) Once the TS interface definitions have been sorted it becomes much more easy to migrate persistences to typescript if desired since much of the typing has already been done by then ;-)

Anyways: I will try to tackle generalization of the db.get/put/del for now first ;-)

**Comment by seriousme** [comment]
Ok, I generalized the encoding/decoding of db.get/put etc

I also tried to use [standard private fields and private methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields) however Standard.js [does not support those yet ](https://github.com/standard/standard/issues/1401) so I had to revert to `_` again :-(

The example in example.js is now also modernized ;-)

Cheers,
Hans

File: persistence.js [hunk 1]
```diff
@@ -22,449 +16,535 @@ const OUTGOING = 'outgoing:'
...
+function endLevel6 (iter) {
+  return new Promise((resolve, reject) => {
+    iter.end(resolve, reject)
+  })
+}
 
-  const trie = this._trie
-  const that = this
+async function * dbValuesLevel6 (db, start, opts) {
+  console.log('level6', typeof db.iterator)
```
Comment (robertsLando) [inline persistence.js]: Remove console.log statement

**Review by robertsLando — CHANGES_REQUESTED** [review]

**Review by mcollina — COMMENTED** [review]
I'm ok in dropping old versions.

File: persistence.js (hunk 1 above)
Comment (seriousme) [reply persistence.js]: Done, see [76642b9](https://github.com/moscajs/aedes-persistence-level/pull/28/commits/76642b967f298865a8fd2bd9007e3f9e49571ac8)

**Comment by robertsLando** [comment]
@seriousme For a cleaner code so, let's drop old Level versions support.

@mcollina What about the usage of generators in place of streams?

**Comment by seriousme** [comment]
Removed support for Level < 8.0.0
Btw: the generators only replace streams internally, the interface to Aedes still supplies Node.js Readable streams ;-)

**Review by robertsLando — APPROVED** [review]

**Comment by robertsLando** [comment]
Thanks @seriousme ! We need help maintaining aedes and all his modules, if you would help us making some other improvements you are welcome! :)

**Pull request merged**